### PR TITLE
Memory manager: introduce virtual memory regions and their calculations

### PR DESCRIPTION
**Pull request by dabekjakub**
Add device tree entry for virtual memory.
Refactor code to make it more readable and split defines into h file.
Add region calculations to share to application.

### CODE REVIEW

**Review by jxstelter — DISMISSED** [review]

**Comment by dcpleung** [comment]
There are compilation errors in the code you submitted.

**Comment by dabekjakub** [comment]
Well thats shamefull xD I must have introduced something while doing "cosmetic" changes. Ty very much.

**Review by nashif — CHANGES_REQUESTED** [review]

File: drivers/mm/mm_drv_intel_adsp_regions.c [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include "mm_drv_intel_adsp.h"
+
+#define APPEND_REGIONS (address, mem_size, attribute)	\
+virtual_memory_regions[i].addr = (void *)(address);	\
+virtual_memory_regions[i].size = mem_size;	\
+virtual_memory_regions[i].attr = attribute;	\
+checksum += mem_size;
+
+struct sys_mm_drv_region
+virtual_memory_regions[CONFIG_MAX_CORE_COUNT + VIRTUAL_REGION_COUNT] = { {0} };
```
[the remark this answers is not in the document]
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp_regions.c:22]: Ohhhh I understand the errors i saw now , Ty :)

File: drivers/mm/mm_drv_intel_adsp.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
```
Comment (pblaszko) [inline drivers/mm/mm_drv_intel_adsp.h:2]: 2022

File: drivers/mm/mm_drv_intel_adsp_regions.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			CORE_HEAP_SIZE, MEM_REG_ATTR_CORE_HEAP);
+	}
+
+	APPEND_REGIONS((uintptr_t)virtual_memory_regions[i - 1].addr + temporary_table[i - 1].size,
+		CORE_HEAP_SIZE, MEM_REG_ATTR_SHARED_HEAP);
+	i++;
+	APPEND_REGIONS((uintptr_t)virtual_memory_regions[i - 1].addr + temporary_table[i - 1].size,
+		CORE_HEAP_SIZE, MEM_REG_ATTR_OPPURTUNISTIC_MEMORY);
+	i++;
+	APPEND_REGIONS(0, 0, 0);
```
Comment (pblaszko) [inline drivers/mm/mm_drv_intel_adsp_regions.c:47]: why APPEND_REGIONS(0, 0, 0)? Please add comment

File: drivers/mm/mm_drv_intel_adsp.h [hunk 4]
```diff
@@ -59,21 +59,33 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 
 /* references counter to physical pages */
 #define UNUSED_L2_START_ALIGNED ROUND_UP(POINTER_TO_UINT(&unused_l2_sram_start_marker), \
 					 CONFIG_MM_DRV_PAGE_SIZE)
 
-
+/* Attribiutes for memory regions */
+#define MEM_REG_ATTR_CORE_HEAP 1U
+#define MEM_REG_ATTR_SHARED_HEAP 2U
+#define MEM_REG_ATTR_OPPURTUNISTIC_MEMORY 4U
```
Comment (pblaszko) [inline drivers/mm/mm_drv_intel_adsp.h:82]: you use OMA_REGION_SIZE and then MEM_REG_ATTR_OPPURTUNISTIC_MEMORY.
For better clarity, use the same name: OMA or OPPORTUNISTIC.
If using OMA, please add comment to definition what does it state for: Opportunistic Memory Allocation

File: drivers/mm/mm_drv_intel_adsp_regions.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp_regions.c:22]: Changed to zephyr config

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 5]
```diff
@@ -581,6 +581,8 @@ static int sys_mm_drv_mm_init(const struct device *dev)
 
 	L2_PHYS_SRAM_REGION.num_blocks = avalible_memory_size / CONFIG_MM_DRV_PAGE_SIZE;
 
+	ret = calculate_memory_regions(UNUSED_L2_START_ALIGNED);
```
[the remark this answers is not in the document]
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:584]: Fixed.

File: drivers/mm/mm_drv_intel_adsp_regions.c (hunk 3 above)
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp_regions.c:47]: Fixed.

File: drivers/mm/mm_drv_intel_adsp.h (hunk 4 above)
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp.h:82]: Fixed.

File: drivers/mm/mm_drv_intel_adsp.h [hunk 6]
```diff
@@ -59,21 +59,33 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 
+/* Define amount of regions other than core heaps that virtual memory will be split to
+ * currently includes shared heap and oma region and one regions set to 0 as for table
+ * iterator end value.
+ */
+#define VIRTUAL_REGION_COUNT 3
+
+#define CORE_HEAP_SIZE 0x100000
+#define SHARED_HEAP_SIZE 0x100000
+#define OPPURTUNISTIC_REGION_SIZE 0x100000
```
Comment (pblaszko) [inline drivers/mm/mm_drv_intel_adsp.h:70]: typo: oppUrtinistic

File: drivers/mm/mm_drv_intel_adsp.h [hunk 7]
```diff
@@ -59,21 +59,33 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 
 /* references counter to physical pages */
 #define UNUSED_L2_START_ALIGNED ROUND_UP(POINTER_TO_UINT(&unused_l2_sram_start_marker), \
 					 CONFIG_MM_DRV_PAGE_SIZE)
 
-
+/* Attribiutes for memory regions */
+#define MEM_REG_ATTR_CORE_HEAP 1U
+#define MEM_REG_ATTR_SHARED_HEAP 2U
+#define MEM_REG_ATTR_OPPURTUNISTIC_MEMORY 4U
```
Comment (pblaszko) [inline drivers/mm/mm_drv_intel_adsp.h:82]: typo: oppUrtinistic

File: drivers/mm/mm_drv_intel_adsp_regions.c [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		virtual_memory_regions[i - 1].addr + temporary_table[i - 1].size),
+		OPPORTUNISTIC_REGION_SIZE, MEM_REG_ATTR_OPPORTUNISTIC_MEMORY, i, &checksum);
+	i++;
+	/* Apending last region as 0 so iterators know where table is over
+	 * check is for size = 0;
+	 */
+	append_region((void *)0, 0, 0, i, &checksum);
+
+	if (checksum > L2_VIRTUAL_SRAM_SIZE)
+		return -EINVAL;
```
Comment (dcpleung) [inline drivers/mm/mm_drv_intel_adsp_regions.c:58]: Brackets are still required for one line statement for if-blocks.

File: drivers/mm/CMakeLists.txt [hunk 9]
```diff
@@ -12,5 +12,6 @@ zephyr_library_sources_ifdef(
 
 zephyr_sources_ifdef(
 	CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+    mm_drv_intel_adsp_regions.c
```
Comment (dcpleung) [inline drivers/mm/CMakeLists.txt:15]: Please follow the indentation in the file.

**Review by dcpleung — CHANGES_REQUESTED** [review]

File: drivers/mm/mm_drv_intel_adsp_regions.c (hunk 8 above)
Comment (dabekjakub) [reply drivers/mm/mm_drv_intel_adsp_regions.c:58]: I did not get any checkpatch issues there - shouldn't we introduce that into it later as an improvment ?

**Review by dcpleung — DISMISSED** [review]

File: drivers/mm/mm_drv_intel_adsp_regions.c (hunk 8 above)
Comment (dcpleung) [reply drivers/mm/mm_drv_intel_adsp_regions.c]: I think checkpatch only warning about unnecessary ones.

**Comment by dabekjakub** [comment]
Had to change position for some defines so they would be visible from application.

**Review by pblaszko — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by mwasko** [comment]
@nashif it looks like your change request is blocking this PR, can you check the updated code?

**Review by mwasko — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**